### PR TITLE
#512 blob storage configuration integration

### PR DESCRIPTION
**Pull request by vivek-devtron**
# Description
Added check for blob storage on ci details, cdDetails and build and deploy page.

Fixes #[ (issue)](https://github.com/devtron-labs/dashboard/issues/512)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] This change requires a documentation update

# How Has This Been Tested?
- verified on build & deploy, there should be any check to ignore cache if blob storage not configured
- verified on ciDetails that if log is not coming and blob storage not configured then show user message to configure
- verified on cdDetails that if log is not coming and blob storage not configured then show user message to configure

# Checklist:

* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [ ] Does this PR require documentation updates?
* [ ] I've updated documentation as required by this PR.
* [x] I have performed a self-review of my own code
* [x] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: src/components/app/details/triggerView/types.ts [hunk 1]
```diff
@@ -83,11 +83,14 @@ export interface CIMaterialProps extends RouteComponentProps<CIMaterialRouterPro
...
     loader: boolean
     setLoader: (isLoading) => void
+    isFirstTrigger?: boolean
+    isCacheAvailable?: boolean
 }
 
 export interface CIMaterialState {
     regexValue: Record<number, { value: string; isInvalid: boolean }>
     selectedCIPipeline?: any
+    isBobStorageConfigured?: boolean
```
Comment (sohel-rp) [inline src/components/app/details/triggerView/types.ts:93]: Seems typo - should be `isBlobStorageConfigured`

File: src/components/app/details/cIDetails/ciDetails.scss [hunk 2]
```diff
@@ -336,4 +336,9 @@
     grid-template-columns: 20px auto;
     gap: 16px;
     place-items: center center;
-}
\ No newline at end of file
+}
+
+.configure-blob-container{
+  position: fixed;
+  bottom: 40px;
```
Comment (sohel-rp) [inline src/components/app/details/cIDetails/ciDetails.scss:343]: can we try using classes for this?

File: src/components/app/config.tsx [hunk 3]
```diff
@@ -18,6 +18,7 @@ export const statusColor = {
     succeeded: '#1dad70',
     running: '#FF7E5B',
     failed: '#f33e3e',
+    error: '#f33e3e',
```
[the remark this answers is not in the document]
Comment (vivek-devtron) [reply src/components/app/config.tsx:21]: We get error from the BE in status field

File: src/components/app/details/triggerView/types.ts (hunk 1 above)
Comment (vivek-devtron) [reply src/components/app/details/triggerView/types.ts:93]: done

File: src/components/app/details/triggerView/TriggerView.tsx [hunk 4]
```diff
@@ -853,6 +854,8 @@ class TriggerView extends Component<TriggerViewProps, TriggerViewState> {
                         getWorkflows={this.getWorkflows}
                         loader={this.state.loader}
                         setLoader={this.setLoader}
+                        isFirstTrigger={nd?.status?.toLowerCase()==='not triggered'}
```
[the remark this answers is not in the document]
Comment (vivek-devtron) [reply src/components/app/details/triggerView/TriggerView.tsx]: Done

File: src/config/constants.ts [hunk 5]
```diff
@@ -187,7 +189,7 @@ export const AppConfigStatus = {
 
 export const FullRoutes = {
     LOGIN: `${Routes.LOGIN}`,
-    CENTRAL: 'https://api.devtron.ai',
+    CENTRAL: 'https://api-stage.devtron.ai',
```
Comment (sohel-rp) [inline src/config/constants.ts:192]: Change API to prod one

File: src/components/app/details/triggerView/ciMaterial.tsx [hunk 6]
```diff
@@ -31,39 +33,96 @@ export class CIMaterial extends Component<CIMaterialProps, CIMaterialState> {
...
+      this.getSecurityModuleStatus()
+    }
+
+    async getSecurityModuleStatus(): Promise<void> {
+      try {
+          const { result } = await getModuleConfigured(ModuleNameMap.BLOB_STORAGE)
+          if (result?.enabled) {
+              this.setState((prevState) => ({
+                  isBlobStorageConfigured: true,
+              }))
```
Comment (sohel-rp) [inline src/components/app/details/triggerView/ciMaterial.tsx:50]: Change to `this.setState({ isBlobStorageConfigured: true })`

File: src/components/app/details/cdDetails/cd.utils.tsx [hunk 7]
```diff
@@ -64,3 +67,39 @@ export function Option(props) {
...
+}
+
+const renderBlobNotConfigured = (): JSX.Element => {
+  return (
+      <>
+          {renderLogsNotAvailable('Logs are available only at runtime.')}
+          <div className="flexbox configure-blob-container pt-8 pr-12 pb-8 pl-12 bcv-1 br-4">
+              <Question className="icon-dim-20 fcv-5" />
+              <span className="fs-13 fw-4 mr-8 ml-8">Want to store logs to view later?</span>
+              <a className="fs-13 fw-6 cb-5 no-decor" href={DOCUMENTATION.ADMIN_PASSWORD} target="_blank">
```
Comment (sohel-rp) [inline src/components/app/details/cdDetails/cd.utils.tsx:90]: Change to `dc__no-decor`

File: src/components/app/details/cIDetails/CIDetails.tsx [hunk 8]
```diff
@@ -991,19 +1015,20 @@ export const LogsRenderer: React.FC<{ triggerDetails: History; setFullScreenView
...
-                )}
-            </div>
-        </>
+
+    return logsNotAvailable && (!isBlobStorageConfigured || !triggerDetails.blobStorageEnabled)  ? (
+      renderConfigurationError(isBlobStorageConfigured)
+    ) : (
+        <div className="logs__body">
+            {logs.map((log, index) => {
+                return <p className="mono fs-14" key={index} dangerouslySetInnerHTML={createMarkup(log)} />
```
Comment (sohel-rp) [inline src/components/app/details/cIDetails/CIDetails.tsx:1024]: Use section prefixes in keys to make them unique.

**Review by sohel-rp — APPROVED** [review]

File: src/components/app/config.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (sohel-rp) [reply src/components/app/config.tsx:21]: Ok

File: src/config/constants.ts (hunk 5 above)
Comment (sohel-rp) [reply src/config/constants.ts:192]: @vivek-devtron As discussed, change this API back to the prod one after testing & before merging.

File: src/components/app/details/cIDetails/CIDetails.tsx [hunk 9]
```diff
@@ -87,16 +90,27 @@ function useCIEventSource(url: string, maxLength?: number) {
...
-        eventSourceRef.current = new EventSource(url, { withCredentials: true })
-        eventSourceRef.current.addEventListener('message', handleMessage)
-        eventSourceRef.current.addEventListener('START_OF_STREAM', handleStreamStart)
-        eventSourceRef.current.addEventListener('END_OF_STREAM', handleStreamEnd)
+        if(url){
+          eventSourceRef.current = new EventSource(url, { withCredentials: true })
+          eventSourceRef.current.addEventListener('message', handleMessage)
+          eventSourceRef.current.addEventListener('START_OF_STREAM', handleStreamStart)
+          eventSourceRef.current.addEventListener('END_OF_STREAM', handleStreamEnd)
+          eventSourceRef.current.addEventListener('error', handleError)
```
Comment (shivani170) [inline src/components/app/details/cIDetails/CIDetails.tsx:108]: Please use constants for 'error', 'message', 'START_OF_STREAM' etc if possible

File: src/components/app/details/cIDetails/CIDetails.tsx [hunk 10]
```diff
@@ -625,9 +646,10 @@ export const TriggerDetails: React.FC<{ triggerDetails: History; abort?: () => P
                         progressing: <ProgressingStatus triggerDetails={triggerDetails} abort={abort} type={type} />,
                         starting: <ProgressingStatus triggerDetails={triggerDetails} abort={abort} type={type} />,
                         failed: <Failed triggerDetails={triggerDetails} type={type} />,
+                        error: <Failed triggerDetails={triggerDetails} type={type} />,
                     }[triggerDetails.status.toLowerCase()]
                 }
-                {!['succeeded', 'healthy', 'running', 'progressing', 'starting', 'failed'].includes(
+                {!['succeeded', 'healthy', 'running', 'progressing', 'starting', 'failed', 'error'].includes(
```
Comment (shivani170) [inline src/components/app/details/cIDetails/CIDetails.tsx:652]: Same here

**Review by shivani170 — APPROVED** [review]

File: src/config/constants.ts [hunk 11]
```diff
@@ -549,4 +551,26 @@ export const ModuleNameMap = {
...
   CICD: 'cicd',
   SECURITY: 'security.clair',
+  BLOB_STORAGE: 'blob-storage',
 }
+
+export const BUILD_STATUS = {
+  NOT_TRIGGERED: 'not triggered'
+}
+
+export const EVENT_STREAM_EVENTS_MAP = {
```
Comment (sohel-rp) [inline src/config/constants.ts:561]: rename this to `EVENTSTREAM_EVENTS_MAP`

**Review by sohel-rp — APPROVED** [review]

**Pull request closed without merging**